extension/fluentbitextension/process.go

```diff
@@ -0,0 +1,207 @@
+	"context"
+	"io"
+	"os"
+	"os/exec"
+	"time"
+
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/component"
+)
+
+type processManager struct {
+	cancel context.CancelFunc
+	conf   *Config
+	logger *zap.Logger
+}
+
+func newProcessManager(conf *Config, logger *zap.Logger) *processManager {
+	return &processManager{
+		conf:   conf,
+		logger: logger,
+	}
+}
+
+type procState string
+
+// A global var that is available only for testing
+var restartDelay = 10 * time.Second
+
+const (
+	Uninitialized procState = "uninitialized"
+	Running       procState = "running"
+	Starting      procState = "starting"
+	ShuttingDown  procState = "shutting-down"
+	Stopped       procState = "stopped"
+	Restarting    procState = "restarting"
+	Errored       procState = "errored"
+)
+
+func constructArgs(tcpEndpoint string) []string {
+	return []string{
+		"--config=/dev/stdin",
+		"--http",
+		"--port=2020",
+		"--flush=1",
+		"-o", "forward://" + tcpEndpoint,
+		"--match=*",
+	}
+}
+
+func (pm *processManager) Start(ctx context.Context, host component.Host) error {
+	childCtx, cancel := context.WithCancel(ctx)
+	pm.cancel = cancel
+
+	args := pm.conf.Args
+	if len(args) == 0 {
+		args = constructArgs(pm.conf.TCPEndpoint)
+	}
+	go run(childCtx, pm.conf.ExecutablePath, args, pm.conf.Config, pm.logger)
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (pm *processManager) Shutdown(ctx context.Context) error {
+	pm.cancel()
+	return nil
+}
+
+func run(ctx context.Context, execPath string, args []string, config string, logger *zap.Logger) {
+	state := Uninitialized
+
+	var cmd *exec.Cmd
+	var err error
+	var stdin io.WriteCloser
+	var stdout io.ReadCloser
+	// procWait is guaranteed to be sent exactly one message per successful process start
+	procWait := make(chan error)
+
+	// A state machine makes the management easier to understand and account
+	// for all of the edge cases when managing a subprocess.
+	for {
+		logger.Debug("Fluent extension changed state", zap.String("state", string(state)))
+
+		switch state {
+		case Errored:
+			logger.Error("FluentBit process died", zap.Error(err))
+			state = Restarting
+
+		case Uninitialized:
+			logger.Debug("Starting fluent subprocess", zap.String("execPath", execPath))
+			cmd, stdin, stdout = createCommand(execPath, args)
+			state = Starting
+		case Starting:
+			err = cmd.Start()
+			if err != nil {
+				state = Errored
+				continue
+			}
+
+			go signalWhenProcessDone(cmd, procWait)
+
+			state = Running
+		case Running:
+			go collectOutput(stdout, logger)
+
+			err = renderConfig(config, stdin)
+			stdin.Close()
+			if err != nil {
+				state = Errored
+				continue
+			}
+
+			select {
+			case err = <-procWait:
+				if ctx.Err() == nil {
+					// We aren't supposed to shutdown yet so this is an error
+					// state.
+					state = Errored
+					continue
+				}
+				state = Stopped
+			case <-ctx.Done():
+				state = ShuttingDown
+			}
+		case ShuttingDown:
+			_ = cmd.Process.Kill()
```

Comment:
It may be worth doing a graceful shutdown by sending cmd.Process.Signal(syscall.SIGTERM) first and waiting for a while. I don't know if fluentbit reacts to SIGTERM as expected though. I would expect it to flush the internal buffers so that we don't loose data.

Reply:
Changed it to a SIGTERM and added a wait in the shutdown for the process to gracefully end (5 seconds max).